lib/puppet/http/service/file_server.rb

```diff
@@ -0,0 +1,103 @@
+require 'puppet/file_serving/metadata'
+
+class Puppet::HTTP::Service::FileServer < Puppet::HTTP::Service
+  API = '/puppet/v3'.freeze
+
+  EXCLUDED_FORMATS = [:yaml, :b64_zlib_yamli, :dot]
+
+  def initialize(client, server, port)
+    url = build_url(API, server || Puppet[:server], port || Puppet[:masterport])
+    super(client, url)
+  end
+
+  def get_file_metadata(mount_point:, path:, environment:,
+                        links: :manage, checksum_type: Puppet[:digest_algorithm], source_permissions: :ignore,
+                        ssl_context: nil)
+
+    model = Puppet::Transaction::Report
+    network_formats = model.supported_formats - EXCLUDED_FORMATS
+    mime_types = network_formats.map { |f| model.get_format(f).mime }
+
+    response = @client.get(
+      with_base_url("/file_metadata/#{mount_point}/#{path}"),
+      headers: add_puppet_headers({ 'ACCEPT' => mime_types.join(', ') }),
+      params: {
+        :links => links,
+        :checksum_type => checksum_type,
+        :source_permissions => source_permissions,
+        :environment => environment
+      },
+      ssl_context: ssl_context
+    )
+
+
+    if response.success?
+      header = response['Content-Type']
+      if header
+        header.gsub!(/\s*;.*$/,'') # strip any charset
+        formatter = Puppet::Network::FormatHandler.mime(header)
+        raise Puppet::HTTP::ProtocolError.new("Content-Type is unsupported") if EXCLUDED_FORMATS.include?(formatter.name)
+      else
+        raise Puppet::HTTP::ProtocolError.new(_("No content type in http response; cannot parse"))
+      end
+      Puppet::FileServing::Metadata.convert_from(formatter.name, response.body.to_s)
```

Comment:
Nit,  I think it's more explicit to call `convert_from` method on the formatter:

```
formatter.convert_from(Puppet::FileServing::Metadata, response.body.to_s)
```

Otherwise, we end up passing the name of the formatter, eg `:json` to the `Metadata#convert_from` method and it has to lookup the formatter we already looked up.

Reply:
```
  1) Puppet::HTTP::Service::FileServer retriving file metadata submits a request for file metadata to the server
     Failure/Error: formatter.convert_from(Puppet::FileServing::Metadata, response.body.to_s)

     NoMethodError:
       undefined method `convert_from' for #<Puppet::Network::Format:0x00007fd6e797f250>
     # ./lib/puppet/http/service/file_server.rb:36:in `get_file_metadata'
     # ./spec/unit/http/service/file_server_spec.rb:63:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:180:in `block (2 levels) in <top (required)>'

  2) Puppet::HTTP::Service::FileServer retriving multiple file metadatas submits a request for file metadata to the server
     Failure/Error: formatter.convert_from_multiple(Puppet::FileServing::Metadata, response.body.to_s)

     NoMethodError:
       undefined method `convert_from_multiple' for #<Puppet::Network::Format:0x00007fd6e797f250>
     # ./lib/puppet/http/service/file_server.rb:60:in `get_file_metadatas'
     # ./spec/unit/http/service/file_server_spec.rb:109:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:180:in `block (2 levels) in <top (required)>'
```

Reply:
Sorry the methods are `intern` and `intern_multiple`. See https://github.com/puppetlabs/puppet/blob/master/lib/puppet/network/format_support.rb#L11-L23. Also we need to handle deserialization errors, so should do something like:

```ruby
begin
  formatter.convert_from(Puppet::FileServing::Metadata, response.body.to_s)
rescue => err
  raise Puppet::HTTP::SerializationError.new("Could not create an instance of Metadata from #{formatter.mime}", err)
end
```

Where we need a new class of error to represent de/serialization errors